src/consensus/dkg.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright 2020 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{network_event::AccumulatingEvent, proof::Proof};
+use crate::{
+    id::{FullId, PublicId},
+    rng::{MainRng, RngCompat},
+    section::EldersInfo,
+    QUORUM_DENOMINATOR, QUORUM_NUMERATOR,
+};
+use bls::{PublicKeySet, SecretKeyShare};
+use bls_dkg::key_gen::{message::Message as DkgMessage, KeyGen};
+use std::{
+    collections::{BTreeMap, BTreeSet, VecDeque},
+    fmt::{self, Debug, Formatter},
+    mem,
+};
+
+/// Generate a BLS SecretKeySet for the given number of participants.
+/// Used for generating first node, or for test.
+pub fn generate_secret_key_set(rng: &mut MainRng, participants: usize) -> bls::SecretKeySet {
+    // The BLS scheme will require more than `participants / 3`
+    // shares in order to construct a full key or signature.
+    let threshold = participants.saturating_sub(1) / 3;
+
+    bls::SecretKeySet::random(threshold, &mut RngCompat(rng))
+}
+
+#[derive(Clone)]
+/// DKG result
+pub struct DkgResult {
+    /// Public key set to verify threshold signatures
+    pub public_key_set: PublicKeySet,
+    /// Secret Key share: None if the node was not participating in the DKG and did not receive
+    /// encrypted shares.
+    pub secret_key_share: Option<SecretKeyShare>,
+}
+
+impl DkgResult {
+    /// Create DkgResult from components
+    pub fn new(public_key_set: PublicKeySet, secret_key_share: Option<SecretKeyShare>) -> Self {
+        Self {
+            public_key_set,
+            secret_key_share,
+        }
+    }
+}
+
+impl Debug for DkgResult {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(
+            formatter,
+            "DkgResult({:?}, {})",
+            self.public_key_set,
+            self.secret_key_share.is_some()
+        )
+    }
+}
+
+/// Returns the number of minumn responsive participants expected for the DKG process
+#[inline]
+pub const fn threshold_count(elder_size: usize) -> usize {
+    // TODO: allow threshold to be configurable.
+    elder_size * QUORUM_NUMERATOR / QUORUM_DENOMINATOR
+}
+
+pub type DkgKey = (BTreeSet<PublicId>, u64);
+
+#[derive(Default)]
```

Comment:
I think this whole module could use more documentation comments. It's not always clear from the method names alone what their purpose is.

Reply:
Ah, right. I will put more comments